br/pkg/lightning/restore/precheck_impl.go

```diff
@@ -0,0 +1,1164 @@
+		}
+	} else {
+		if err := taskMgr.CheckTasksExclusively(ctx, func(tasks []taskMeta) ([]taskMeta, error) {
+			clusterAvail = 0
+			clusterSource = 0
+			restoreStarted := false
+			for _, task := range tasks {
+				if task.status > taskMetaStatusInitial {
+					restoreStarted = true
+				}
+				clusterSource += task.sourceBytes
+				if task.clusterAvail > 0 {
+					clusterAvail = task.clusterAvail
+				}
+			}
+			if restoreStarted || clusterAvail > 0 {
+				return nil, nil
+			}
+
+			clusterAvail, err = ci.getClusterAvail(ctx)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			newTasks := append([]taskMeta(nil), tasks...)
+			for i := 0; i < len(newTasks); i++ {
+				newTasks[i].clusterAvail = clusterAvail
+			}
+			return newTasks, nil
+		}); err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	replicaCount, err := ci.getReplicaCount(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	estimateSize := clusterSource * replicaCount
+	if estimateSize > clusterAvail {
+		theResult.Passed = false
+		theResult.Message = fmt.Sprintf("Cluster doesn't have enough space, available is %s, but we need %s",
+			units.BytesSize(float64(clusterAvail)), units.BytesSize(float64(estimateSize)))
+	} else {
+		theResult.Message = fmt.Sprintf("Cluster available is rich, available is %s, we need %s",
+			units.BytesSize(float64(clusterAvail)), units.BytesSize(float64(estimateSize)))
+	}
+	return theResult, nil
+}
+
+type clusterVersionCheckItem struct {
+	preInfoGetter PreRestoreInfoGetter
+	dbMetas       []*mydump.MDDatabaseMeta
+}
+
+func NewClusterVersionCheckItem(preInfoGetter PreRestoreInfoGetter, dbMetas []*mydump.MDDatabaseMeta) PrecheckItem {
+	return &clusterVersionCheckItem{
+		preInfoGetter: preInfoGetter,
+		dbMetas:       dbMetas,
+	}
+}
+
+func (ci *clusterVersionCheckItem) GetCheckItemID() CheckItemID {
+	return CheckTargetClusterVersion
+}
+
+func (ci *clusterVersionCheckItem) Check(ctx context.Context) (*CheckResult, error) {
+	theResult := &CheckResult{
+		Item:     ci.GetCheckItemID(),
+		Severity: Critical,
+		Passed:   true,
+		Message:  "Cluster version check passed",
+	}
+	checkCtx := WithPreInfoGetterDBMetas(ctx, ci.dbMetas)
+	if err := ci.preInfoGetter.CheckVersionRequirements(checkCtx); err != nil {
+		err := common.NormalizeError(err)
+		theResult.Passed = false
+		theResult.Message = fmt.Sprintf("Cluster version check failed: %s", err.Error())
+	}
+	return theResult, nil
+}
+
+type emptyRegionCheckItem struct {
+	preInfoGetter PreRestoreInfoGetter
+	dbMetas       []*mydump.MDDatabaseMeta
+}
+
+func NewEmptyRegionCheckItem(preInfoGetter PreRestoreInfoGetter, dbMetas []*mydump.MDDatabaseMeta) PrecheckItem {
+	return &emptyRegionCheckItem{
+		preInfoGetter: preInfoGetter,
+		dbMetas:       dbMetas,
+	}
+}
+
+func (ci *emptyRegionCheckItem) GetCheckItemID() CheckItemID {
+	return CheckTargetClusterEmptyRegion
+}
+
+func (ci *emptyRegionCheckItem) Check(ctx context.Context) (*CheckResult, error) {
```

Comment:
I remember we are considering removing this check (in both Lightning and import service)?